### ISSUE TITLE
Crypto test is leaking resources

### ISSUE THREAD

**Issue by gyuri-szing** [ARM-software/psa-arch-tests#350]
**Summary**
The crypto tests seem to fail cleaning up test data and repeated execution makes test cases fail after a few iterations. Also this affects other non psa-arch-test tests depending on free space available in psa-its.
In real world deployments, where psa-its uses persistent storage, running the crypto test will alter the device state persistently. Removing the left-over data may be complicated as psa storage can not list data, and working out the assets handles created by other clients is difficult.

**Severity**
As we are not using any device with persistent storage,  and the only effect is our test become order dependent. We have to execute  psa tests last.

**Environment**
I am running the tests on the `quemuarm64-secureboot` yocto platform, and I am testing the psa-crypto SP developed in the Trusted Services project. I think the issue could be recreated on any platform. TS service tests do not fail even running for an extended period of time, thus I think the problem is not in the psa-crypto or psa-its implementation.

I am running the following code in the NWd shell:
```
root@qemuarm64-secureboot:~#  while true; do psa-crypto-api-test | tee test.log | grep "TOTAL FAILED    : 2" || break; sleep 3; done
TOTAL FAILED    : 2
TOTAL FAILED    : 2
TOTAL FAILED    : 2
TOTAL FAILED    : 2
TOTAL FAILED    : 2
TOTAL FAILED    : 2
TOTAL FAILED    : 2
TOTAL FAILED    : 2
root@qemuarm64-secureboot:~#
```
We have two unsupported tests, hence looking for failures increase from 2. When this happens test 218 starts to fail like this:
```
TEST: 218 | DESCRIPTION: Testing crypto key derivation APIs | UT: psa_key_derivation_input_key
[Info] Executing tests from non-secure
[Check 1] Test psa_key_derivation_input_key 16 Byte Key
[Check 2] Test psa_key_derivation_input_key with invalid usage
[Check 3] Test psa_key_derivation_input_key with step as label
[Check 4] Test psa_key_derivation_input_key with step as info
[Check 5] Test psa_key_derivation_input_key with step as seed
[Check 6] Test psa_key_derivation_input_key with step as salt
[Check 7] Test psa_key_derivation_input_key with key type as AES(not derive)
[Check 8] Test psa_key_derivation_input_key incorrect key algorithm
[Check 9] Test psa_key_derivation_input_key with key type as 2048 RSA public key
[Check 10] Test psa_key_derivation_input_key with key type as RSA 2048 keypair
	Failed at Checkpoint: 3
	Actual: -141
	Expected: 0

TEST RESULT: FAILED (Error Code=0x1)
```
If I continue running the test by "removing" the exit condition, number of failures shows further increase from 3 (I did not saved the test output as I think it is irrelevant which test cases start to fail too):
```
root@qemuarm64-secureboot:~#  while true; do psa-crypto-api-test | tee test.log | grep "TOTAL FAILED    :" || break; sleep 1; done
TOTAL FAILED    : 5
TOTAL FAILED    : 7
TOTAL FAILED    : 8
...
```
Number of failures seems to flat out at 8.

**Comment by athoelke** [comment]
Status code -141, reported within the test itself, is [`PSA_ERROR_INSUFFICIENT_MEMORY`](https://arm-software.github.io/psa-api/status-code/1.0/api/status-codes.html#c.PSA_ERROR_INSUFFICIENT_MEMORY).

This should be related to issues with allocating volatile memory (RAM) for an operation, not with NVM, which is meant to be reported using [`PSA_ERROR_INSUFFICIENT_STORAGE`](https://arm-software.github.io/psa-api/status-code/1.0/api/status-codes.html#c.PSA_ERROR_INSUFFICIENT_STORAGE).

Services can use `PSA_ERROR_INSUFFICIENT_MEMORY` to also report exhaustion of other in-memory resources, which are not dynamically expanded at runtime. So there might be an issue with incomplete clean-up of volatile keys, not necessarily just persistent keys (that would use ITS for storage).

**Comment by gyuri-szing** [comment]
> Services can use PSA_ERROR_INSUFFICIENT_MEMORY to also report exhaustion of other in-memory resources, which are not dynamically expanded at runtime. So there might be an issue with incomplete clean-up of volatile keys, not necessarily just persistent keys (that would use ITS for storage).

AFAIS the error value is returned by MbedTLS. I need some time to investigate the call path and to see what is going on. As I mentioned, other tests are affected by the psa-crypto tests, so definitely some leftover data is left behind. Will come back with more details in a few days.

**Comment by jk-arm** [comment]
closing this as of now, lets reopen when required.

**Comment by adeaarm** [comment]
Just noticed this task here and I am not sure if any additional investigation has been done there since the last message, but from experience that error code gets returned when the slots for the keys are exhausted, which is probably confirming that some tests might not clean up properly the keys they have imported.

**Comment by gyuri-szing** [comment]
Hi Antonio, unfortunately I did not had the time to look into this more. I am not sure if the error still present as most components have been updated multiple times since then. Investigating this is still on my plate although with low priority.

If this is a psa-acs problem, then other environments should likely be affected too. (e.g. when testing MbedTLS in linux user-space or when testing TF-M.)
